// ng-zorro-antd/esm5/core/testing/fake-viewport-ruler.js
/**
 * @fileoverview added by tsickle
 * Generated from: fake-viewport-ruler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @record
 */
function FakeViewportRect() { }
if (false) {
    /** @type {?} */
    FakeViewportRect.prototype.left;
    /** @type {?} */
    FakeViewportRect.prototype.top;
    /** @type {?} */
    FakeViewportRect.prototype.width;
    /** @type {?} */
    FakeViewportRect.prototype.height;
    /** @type {?} */
    FakeViewportRect.prototype.bottom;
    /** @type {?} */
    FakeViewportRect.prototype.right;
}
/**
 * @record
 */
function FakeViewportSize() { }
if (false) {
    /** @type {?} */
    FakeViewportSize.prototype.width;
    /** @type {?} */
    FakeViewportSize.prototype.height;
}
/**
 * @record
 */
function FakeViewportScrollPosition() { }
if (false) {
    /** @type {?} */
    FakeViewportScrollPosition.prototype.top;
    /** @type {?} */
    FakeViewportScrollPosition.prototype.left;
}
/**
 * \@docs-private
 */
var /**
 * \@docs-private
 */
FakeViewportRuler = /** @class */ (function () {
    function FakeViewportRuler() {
    }
    /**
     * @return {?}
     */
    FakeViewportRuler.prototype.getViewportRect = /**
     * @return {?}
     */
    function () {
        return {
            left: 0,
            top: 0,
            width: 1014,
            height: 686,
            bottom: 686,
            right: 1014
        };
    };
    /**
     * @return {?}
     */
    FakeViewportRuler.prototype.getViewportSize = /**
     * @return {?}
     */
    function () {
        return { width: 1014, height: 686 };
    };
    /**
     * @return {?}
     */
    FakeViewportRuler.prototype.getViewportScrollPosition = /**
     * @return {?}
     */
    function () {
        return { top: 0, left: 0 };
    };
    return FakeViewportRuler;
}());
/**
 * \@docs-private
 */
export { FakeViewportRuler };
