// ngx-validators/esm5/components/creditcard/creditcard.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { CreditCardValidators } from './creditcard-validators';
var CreditCardValidatorDirective = /** @class */ (function () {
    function CreditCardValidatorDirective() {
        this.creditCard = 'all';
    }
    /**
     * @return {?}
     */
    CreditCardValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        switch (this.creditCard) {
            case 'all':
                this.validator = CreditCardValidators.isCreditCard;
                break;
            case 'americanExpress':
                this.validator = CreditCardValidators.americanExpress;
                break;
            case 'dinersclub':
                this.validator = CreditCardValidators.dinersclub;
                break;
            case 'discover':
                this.validator = CreditCardValidators.discover;
                break;
            case 'jcb':
                this.validator = CreditCardValidators.jcb;
                break;
            case 'maestro':
                this.validator = CreditCardValidators.maestro;
                break;
            case 'mastercard':
                this.validator = CreditCardValidators.mastercard;
                break;
            case 'visa':
                this.validator = CreditCardValidators.visa;
                break;
            default:
                this.validator = CreditCardValidators.isCreditCard;
                break;
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    CreditCardValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    CreditCardValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[creditCard][formControlName],[creditCard][formControl],[creditCard][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return CreditCardValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    CreditCardValidatorDirective.propDecorators = {
        creditCard: [{ type: Input }]
    };
    return CreditCardValidatorDirective;
}());
export { CreditCardValidatorDirective };
if (false) {
    /** @type {?} */
    CreditCardValidatorDirective.prototype.creditCard;
    /** @type {?} */
    CreditCardValidatorDirective.prototype.validator;
}
