// ngx-validators/esm2015/components/email/email.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { EmailValidators } from './email-validators';
export class EmailValidatorDirective {
    constructor() {
        this.email = 'normal';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        switch (this.email) {
            case 'simple':
                this.validator = EmailValidators.simple;
                break;
            case 'normal':
                this.validator = EmailValidators.normal;
                break;
            default:
                this.validator = EmailValidators.normal;
                break;
        }
    }
    /**
     * @param {?} c
     * @return {?}
     */
    validate(c) {
        return this.validator(c);
    }
}
EmailValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[email][formControlName],[email][formControl],[email][ngModel]',
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line:no-forward-ref
                        useExisting: forwardRef(() => EmailValidatorDirective),
                        multi: true
                    }]
            },] }
];
EmailValidatorDirective.propDecorators = {
    email: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    EmailValidatorDirective.prototype.email;
    /** @type {?} */
    EmailValidatorDirective.prototype.validator;
}
export class EmailSuggestValidatorDirective {
    /**
     * @return {?}
     */
    ngOnInit() {
        this.validator = EmailValidators.suggest(this.emailSuggest);
    }
    /**
     * @param {?} c
     * @return {?}
     */
    validate(c) {
        return this.validator(c);
    }
}
EmailSuggestValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[emailSuggest][formControlName],[emailSuggest][formControl],[emailSuggest][ngModel]',
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line:no-forward-ref
                        useExisting: forwardRef(() => EmailSuggestValidatorDirective),
                        multi: true
                    }]
            },] }
];
EmailSuggestValidatorDirective.propDecorators = {
    emailSuggest: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    EmailSuggestValidatorDirective.prototype.emailSuggest;
    /** @type {?} */
    EmailSuggestValidatorDirective.prototype.validator;
}
