// ng-zorro-antd/esm2015/modal/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
export { ModalOptions } from './modal-types';
export { NzModalService } from './modal.service';
export { NzModalRef } from './modal-ref';
export { NZ_MODAL_CONFIG } from './modal-config';
export { NzModalComponent } from './modal.component';
export { NzModalFooterDirective } from './modal-footer.directive';
export { NzModalModule } from './modal.module';
export { NzModalConfirmContainerComponent } from './modal-confirm-container.component';
export { NzModalContainerComponent } from './modal-container.component';
export { throwNzModalContentAlreadyAttachedError, BaseModalContainer } from './modal-container';
export { NzModalCloseComponent } from './modal-close.component';
export { NzModalTitleComponent } from './modal-title.component';
export { NzModalLegacyAPI } from './modal-legacy-api';
export {} from './modal-types';
export { nzModalAnimations } from './modal-animations';
export { NzModalFooterComponent } from './modal-footer.component';
export { applyConfigDefaults, setContentInstanceParams, getConfigFromComponent } from './utils';
