// ngx-validators/esm5/components/email/email.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { EmailValidators } from './email-validators';
var EmailValidatorDirective = /** @class */ (function () {
    function EmailValidatorDirective() {
        this.email = 'normal';
    }
    /**
     * @return {?}
     */
    EmailValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        switch (this.email) {
            case 'simple':
                this.validator = EmailValidators.simple;
                break;
            case 'normal':
                this.validator = EmailValidators.normal;
                break;
            default:
                this.validator = EmailValidators.normal;
                break;
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    EmailValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    EmailValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[email][formControlName],[email][formControl],[email][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return EmailValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    EmailValidatorDirective.propDecorators = {
        email: [{ type: Input }]
    };
    return EmailValidatorDirective;
}());
export { EmailValidatorDirective };
if (false) {
    /** @type {?} */
    EmailValidatorDirective.prototype.email;
    /** @type {?} */
    EmailValidatorDirective.prototype.validator;
}
var EmailSuggestValidatorDirective = /** @class */ (function () {
    function EmailSuggestValidatorDirective() {
    }
    /**
     * @return {?}
     */
    EmailSuggestValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = EmailValidators.suggest(this.emailSuggest);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    EmailSuggestValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    EmailSuggestValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[emailSuggest][formControlName],[emailSuggest][formControl],[emailSuggest][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return EmailSuggestValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    EmailSuggestValidatorDirective.propDecorators = {
        emailSuggest: [{ type: Input }]
    };
    return EmailSuggestValidatorDirective;
}());
export { EmailSuggestValidatorDirective };
if (false) {
    /** @type {?} */
    EmailSuggestValidatorDirective.prototype.emailSuggest;
    /** @type {?} */
    EmailSuggestValidatorDirective.prototype.validator;
}
