// ng-zorro-antd/bundles/ng-zorro-antd-button.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ng-zorro-antd/core/config'), require('ng-zorro-antd/core/util'), require('ng-zorro-antd/icon'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('ng-zorro-antd/core/transition-patch'), require('ng-zorro-antd/core/wave')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/button', ['exports', '@angular/core', 'ng-zorro-antd/core/config', 'ng-zorro-antd/core/util', 'ng-zorro-antd/icon', 'rxjs', 'rxjs/operators', '@angular/common', 'ng-zorro-antd/core/transition-patch', 'ng-zorro-antd/core/wave'], factory) :
    (global = global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd'].button = {}), global.ng.core, global['ng-zorro-antd'].core.config, global['ng-zorro-antd'].core.util, global['ng-zorro-antd'].icon, global.rxjs, global.rxjs.operators, global.ng.common, global['ng-zorro-antd'].core['transition-patch'], global['ng-zorro-antd'].core.wave));
}(this, (function (exports, core, config, util, icon, rxjs, operators, common, transitionPatch, wave) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: button.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NZ_CONFIG_COMPONENT_NAME = 'button';
    var NzButtonComponent = /** @class */ (function () {
        function NzButtonComponent(elementRef, cdr, renderer, nzConfigService) {
            var _this = this;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.renderer = renderer;
            this.nzConfigService = nzConfigService;
            this.nzBlock = false;
            this.nzGhost = false;
            this.nzSearch = false;
            this.nzLoading = false;
            this.nzDanger = false;
            this.nzType = null;
            this.nzShape = null;
            this.destroy$ = new rxjs.Subject();
            this.loading$ = new rxjs.Subject();
            this.nzConfigService
                .getConfigChangeEventForComponent(NZ_CONFIG_COMPONENT_NAME)
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe((/**
             * @return {?}
             */
            function () {
                _this.cdr.markForCheck();
            }));
        }
        /**
         * @param {?} nodes
         * @param {?} renderer
         * @return {?}
         */
        NzButtonComponent.prototype.insertSpan = /**
         * @param {?} nodes
         * @param {?} renderer
         * @return {?}
         */
        function (nodes, renderer) {
            nodes.forEach((/**
             * @param {?} node
             * @return {?}
             */
            function (node) {
                if (node.nodeName === '#text') {
                    /** @type {?} */
                    var span = renderer.createElement('span');
                    /** @type {?} */
                    var parent_1 = renderer.parentNode(node);
                    renderer.insertBefore(parent_1, span, node);
                    renderer.appendChild(span, node);
                }
            }));
        };
        /**
         * @param {?} element
         * @param {?} renderer
         * @return {?}
         */
        NzButtonComponent.prototype.assertIconOnly = /**
         * @param {?} element
         * @param {?} renderer
         * @return {?}
         */
        function (element, renderer) {
            /** @type {?} */
            var listOfNode = Array.from(element.childNodes);
            /** @type {?} */
            var iconCount = listOfNode.filter((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.nodeName === 'I'; })).length;
            /** @type {?} */
            var noText = listOfNode.every((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.nodeName !== '#text'; }));
            /** @type {?} */
            var noSpan = listOfNode.every((/**
             * @param {?} node
             * @return {?}
             */
            function (node) { return node.nodeName !== 'SPAN'; }));
            /** @type {?} */
            var isIconOnly = noSpan && noText && iconCount === 1;
            if (isIconOnly) {
                renderer.addClass(element, 'ant-btn-icon-only');
            }
        };
        /**
         * @param {?} changes
         * @return {?}
         */
        NzButtonComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
        function (changes) {
            var nzLoading = changes.nzLoading;
            if (nzLoading) {
                this.loading$.next(this.nzLoading);
            }
        };
        /**
         * @return {?}
         */
        NzButtonComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            this.assertIconOnly(this.elementRef.nativeElement, this.renderer);
            this.insertSpan(this.elementRef.nativeElement.childNodes, this.renderer);
        };
        /**
         * @return {?}
         */
        NzButtonComponent.prototype.ngAfterContentInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.loading$
                .pipe(operators.startWith(this.nzLoading), operators.filter((/**
             * @return {?}
             */
            function () { return !!_this.nzIconDirectiveElement; })), operators.takeUntil(this.destroy$))
                .subscribe((/**
             * @param {?} loading
             * @return {?}
             */
            function (loading) {
                /** @type {?} */
                var nativeElement = _this.nzIconDirectiveElement.nativeElement;
                if (loading) {
                    _this.renderer.setStyle(nativeElement, 'display', 'none');
                }
                else {
                    _this.renderer.removeStyle(nativeElement, 'display');
                }
            }));
        };
        /**
         * @return {?}
         */
        NzButtonComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.destroy$.next();
            this.destroy$.complete();
        };
        NzButtonComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'button[nz-button], a[nz-button]',
                        exportAs: 'nzButton',
                        preserveWhitespaces: false,
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        encapsulation: core.ViewEncapsulation.None,
                        template: "\n    <i nz-icon nzType=\"loading\" *ngIf=\"nzLoading\"></i>\n    <ng-content></ng-content>\n  ",
                        host: {
                            '[class.ant-btn]': "true",
                            '[class.ant-btn-primary]': "nzType === 'primary'",
                            '[class.ant-btn-dashed]': "nzType === 'dashed'",
                            '[class.ant-btn-link]': "nzType === 'link'",
                            '[class.ant-btn-danger]': "nzType === 'danger'",
                            '[class.ant-btn-circle]': "nzShape === 'circle'",
                            '[class.ant-btn-round]': "nzShape === 'round'",
                            '[class.ant-btn-lg]': "nzSize === 'large'",
                            '[class.ant-btn-sm]': "nzSize === 'small'",
                            '[class.ant-btn-dangerous]': "nzDanger",
                            '[class.ant-btn-loading]': "nzLoading",
                            '[class.ant-btn-background-ghost]': "nzGhost",
                            '[class.ant-btn-block]': "nzBlock",
                            '[class.ant-input-search-button]': "nzSearch"
                        }
                    }] }
        ];
        /** @nocollapse */
        NzButtonComponent.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.ChangeDetectorRef },
            { type: core.Renderer2 },
            { type: config.NzConfigService }
        ]; };
        NzButtonComponent.propDecorators = {
            nzIconDirectiveElement: [{ type: core.ContentChild, args: [icon.NzIconDirective, { read: core.ElementRef },] }],
            nzBlock: [{ type: core.Input }],
            nzGhost: [{ type: core.Input }],
            nzSearch: [{ type: core.Input }],
            nzLoading: [{ type: core.Input }],
            nzDanger: [{ type: core.Input }],
            nzType: [{ type: core.Input }],
            nzShape: [{ type: core.Input }],
            nzSize: [{ type: core.Input }]
        };
        __decorate([
            util.InputBoolean(),
            __metadata("design:type", Boolean)
        ], NzButtonComponent.prototype, "nzBlock", void 0);
        __decorate([
            util.InputBoolean(),
            __metadata("design:type", Boolean)
        ], NzButtonComponent.prototype, "nzGhost", void 0);
        __decorate([
            util.InputBoolean(),
            __metadata("design:type", Boolean)
        ], NzButtonComponent.prototype, "nzSearch", void 0);
        __decorate([
            util.InputBoolean(),
            __metadata("design:type", Boolean)
        ], NzButtonComponent.prototype, "nzLoading", void 0);
        __decorate([
            util.InputBoolean(),
            __metadata("design:type", Boolean)
        ], NzButtonComponent.prototype, "nzDanger", void 0);
        __decorate([
            config.WithConfig(NZ_CONFIG_COMPONENT_NAME, 'default'),
            __metadata("design:type", String)
        ], NzButtonComponent.prototype, "nzSize", void 0);
        return NzButtonComponent;
    }());
    if (false) {
        /** @type {?} */
        NzButtonComponent.ngAcceptInputType_nzBlock;
        /** @type {?} */
        NzButtonComponent.ngAcceptInputType_nzGhost;
        /** @type {?} */
        NzButtonComponent.ngAcceptInputType_nzSearch;
        /** @type {?} */
        NzButtonComponent.ngAcceptInputType_nzLoading;
        /** @type {?} */
        NzButtonComponent.ngAcceptInputType_nzDanger;
        /** @type {?} */
        NzButtonComponent.prototype.nzIconDirectiveElement;
        /** @type {?} */
        NzButtonComponent.prototype.nzBlock;
        /** @type {?} */
        NzButtonComponent.prototype.nzGhost;
        /** @type {?} */
        NzButtonComponent.prototype.nzSearch;
        /** @type {?} */
        NzButtonComponent.prototype.nzLoading;
        /** @type {?} */
        NzButtonComponent.prototype.nzDanger;
        /** @type {?} */
        NzButtonComponent.prototype.nzType;
        /** @type {?} */
        NzButtonComponent.prototype.nzShape;
        /** @type {?} */
        NzButtonComponent.prototype.nzSize;
        /**
         * @type {?}
         * @private
         */
        NzButtonComponent.prototype.destroy$;
        /**
         * @type {?}
         * @private
         */
        NzButtonComponent.prototype.loading$;
        /**
         * @type {?}
         * @private
         */
        NzButtonComponent.prototype.elementRef;
        /**
         * @type {?}
         * @private
         */
        NzButtonComponent.prototype.cdr;
        /**
         * @type {?}
         * @private
         */
        NzButtonComponent.prototype.renderer;
        /** @type {?} */
        NzButtonComponent.prototype.nzConfigService;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: button-group.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzButtonGroupComponent = /** @class */ (function () {
        function NzButtonGroupComponent() {
            this.nzSize = 'default';
        }
        NzButtonGroupComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'nz-button-group',
                        exportAs: 'nzButtonGroup',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        encapsulation: core.ViewEncapsulation.None,
                        host: {
                            '[class.ant-btn-group]': "true",
                            '[class.ant-btn-group-lg]': "nzSize === 'large'",
                            '[class.ant-btn-group-sm]': "nzSize === 'small'"
                        },
                        preserveWhitespaces: false,
                        template: "\n    <ng-content></ng-content>\n  "
                    }] }
        ];
        NzButtonGroupComponent.propDecorators = {
            nzSize: [{ type: core.Input }]
        };
        return NzButtonGroupComponent;
    }());
    if (false) {
        /** @type {?} */
        NzButtonGroupComponent.prototype.nzSize;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: button.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzButtonModule = /** @class */ (function () {
        function NzButtonModule() {
        }
        NzButtonModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [NzButtonComponent, NzButtonGroupComponent],
                        exports: [NzButtonComponent, NzButtonGroupComponent, transitionPatch.ɵNzTransitionPatchModule, wave.NzWaveModule],
                        imports: [common.CommonModule, wave.NzWaveModule, icon.NzIconModule, transitionPatch.ɵNzTransitionPatchModule]
                    },] }
        ];
        return NzButtonModule;
    }());

    exports.NzButtonComponent = NzButtonComponent;
    exports.NzButtonGroupComponent = NzButtonGroupComponent;
    exports.NzButtonModule = NzButtonModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
