// ng-zorro-antd/esm5/core/tree/nz-tree-base-util.js
/**
 * @fileoverview added by tsickle
 * Generated from: nz-tree-base-util.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __read, __spread } from "tslib";
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
/**
 * @param {?} node
 * @return {?}
 */
export function isCheckDisabled(node) {
    var isDisabled = node.isDisabled, isDisableCheckbox = node.isDisableCheckbox;
    return !!(isDisabled || isDisableCheckbox);
}
/**
 * @param {?} needle
 * @param {?} haystack
 * @return {?}
 */
export function isInArray(needle, haystack) {
    return haystack.length > 0 && haystack.indexOf(needle) > -1;
}
/**
 * @param {?} level
 * @param {?} index
 * @return {?}
 */
export function getPosition(level, index) {
    return level + "-" + index;
}
/**
 * @param {?} key
 * @param {?} pos
 * @return {?}
 */
export function getKey(key, pos) {
    if (key !== null && key !== undefined) {
        return key;
    }
    return pos;
}
/**
 * Flat nest tree data into flatten list. This is used for virtual list render.
 * @param {?=} treeNodeList Origin data node list
 * @param {?=} expandedKeys
 * need expanded keys, provides `true` means all expanded (used in `rc-tree-select`).
 * @return {?}
 */
export function flattenTreeData(treeNodeList, expandedKeys) {
    if (treeNodeList === void 0) { treeNodeList = []; }
    if (expandedKeys === void 0) { expandedKeys = []; }
    /** @type {?} */
    var expandedKeySet = new Set(expandedKeys === true ? [] : expandedKeys);
    /** @type {?} */
    var flattenList = [];
    /**
     * @param {?} list
     * @param {?=} parent
     * @return {?}
     */
    function dig(list, parent) {
        if (parent === void 0) { parent = null; }
        return list.map((/**
         * @param {?} treeNode
         * @param {?} index
         * @return {?}
         */
        function (treeNode, index) {
            /** @type {?} */
            var pos = getPosition(parent ? parent.pos : '0', index);
            /** @type {?} */
            var mergedKey = getKey(treeNode.key, pos);
            treeNode.isStart = __spread((parent ? parent.isStart : []), [index === 0]);
            treeNode.isEnd = __spread((parent ? parent.isEnd : []), [index === list.length - 1]);
            // Add FlattenDataNode into list
            // TODO: only need data here.
            /** @type {?} */
            var flattenNode = {
                parent: parent,
                pos: pos,
                children: [],
                data: treeNode,
                isStart: __spread((parent ? parent.isStart : []), [index === 0]),
                isEnd: __spread((parent ? parent.isEnd : []), [index === list.length - 1])
            };
            flattenList.push(flattenNode);
            // Loop treeNode children
            if (expandedKeys === true || expandedKeySet.has(mergedKey) || treeNode.isExpanded) {
                flattenNode.children = dig(treeNode.children || [], flattenNode);
            }
            else {
                flattenNode.children = [];
            }
            return flattenNode;
        }));
    }
    dig(treeNodeList);
    return flattenList;
}
