// ng-zorro-antd/__ivy_ngcc__/fesm5/ng-zorro-antd-core-element-patch.js
import { Directive, ElementRef, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * Generated from: element-patch.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * A patch directive to select the element of a component.
 */
import * as ɵngcc0 from '@angular/core';
var NzElementPatchDirective = /** @class */ (function () {
    function NzElementPatchDirective(elementRef) {
        this.elementRef = elementRef;
    }
    /** @nocollapse */
    NzElementPatchDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
NzElementPatchDirective.ɵfac = function NzElementPatchDirective_Factory(t) { return new (t || NzElementPatchDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NzElementPatchDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NzElementPatchDirective, selectors: [["", "nz-element", ""]], exportAs: ["nzElement"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzElementPatchDirective, [{
        type: Directive,
        args: [{
                selector: '[nz-element]',
                exportAs: 'nzElement'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
    return NzElementPatchDirective;
}());
if (false) {
    /** @type {?} */
    NzElementPatchDirective.prototype.elementRef;
}

/**
 * @fileoverview added by tsickle
 * Generated from: element-patch.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NzElementPatchModule = /** @class */ (function () {
    function NzElementPatchModule() {
    }
NzElementPatchModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzElementPatchModule });
NzElementPatchModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzElementPatchModule_Factory(t) { return new (t || NzElementPatchModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzElementPatchModule, { declarations: function () { return [NzElementPatchDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NzElementPatchDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzElementPatchModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NzElementPatchDirective],
                exports: [NzElementPatchDirective]
            }]
    }], function () { return []; }, null); })();
    return NzElementPatchModule;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ng-zorro-antd-core-element-patch.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { NzElementPatchDirective, NzElementPatchModule };
