// ng-zorro-antd/esm5/slider/typings.js
/**
 * @fileoverview added by tsickle
 * Generated from: typings.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
/**
 * @record
 */
export function NzMarkObj() { }
if (false) {
    /** @type {?|undefined} */
    NzMarkObj.prototype.style;
    /** @type {?} */
    NzMarkObj.prototype.label;
}
var NzMarks = /** @class */ (function () {
    function NzMarks() {
    }
    return NzMarks;
}());
export { NzMarks };
/**
 * Processed steps that would be passed to sub components.
 * @record
 */
export function NzExtendedMark() { }
if (false) {
    /** @type {?} */
    NzExtendedMark.prototype.value;
    /** @type {?} */
    NzExtendedMark.prototype.offset;
    /** @type {?} */
    NzExtendedMark.prototype.config;
}
/**
 * Marks that would be rendered.
 * @record
 */
export function NzDisplayedMark() { }
if (false) {
    /** @type {?} */
    NzDisplayedMark.prototype.active;
    /** @type {?} */
    NzDisplayedMark.prototype.label;
    /** @type {?|undefined} */
    NzDisplayedMark.prototype.style;
}
/**
 * Steps that would be rendered.
 * @record
 */
export function NzDisplayedStep() { }
if (false) {
    /** @type {?} */
    NzDisplayedStep.prototype.active;
    /** @type {?|undefined} */
    NzDisplayedStep.prototype.style;
}
/**
 * @record
 */
export function NzSliderHandler() { }
if (false) {
    /** @type {?} */
    NzSliderHandler.prototype.offset;
    /** @type {?} */
    NzSliderHandler.prototype.value;
    /** @type {?} */
    NzSliderHandler.prototype.active;
}
