// ngx-validators/esm5/components/universal/universal.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { UniversalValidators } from './universal-validators';
var WhiteSpaceValidatorDirective = /** @class */ (function () {
    function WhiteSpaceValidatorDirective() {
    }
    /**
     * @return {?}
     */
    WhiteSpaceValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.noWhitespace;
    };
    /**
     * @param {?} c
     * @return {?}
     */
    WhiteSpaceValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    WhiteSpaceValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[noWhitespace][formControlName],[noWhitespace][formControl],[noWhitespace][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return WhiteSpaceValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    return WhiteSpaceValidatorDirective;
}());
export { WhiteSpaceValidatorDirective };
if (false) {
    /** @type {?} */
    WhiteSpaceValidatorDirective.prototype.validator;
}
var EmptyStringValidatorDirective = /** @class */ (function () {
    function EmptyStringValidatorDirective() {
    }
    /**
     * @return {?}
     */
    EmptyStringValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.noEmptyString;
    };
    /**
     * @param {?} c
     * @return {?}
     */
    EmptyStringValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    EmptyStringValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[noEmptyString][formControlName],[noEmptyString][formControl],[noEmptyString][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return EmptyStringValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    return EmptyStringValidatorDirective;
}());
export { EmptyStringValidatorDirective };
if (false) {
    /** @type {?} */
    EmptyStringValidatorDirective.prototype.validator;
}
var IsNumberValidatorDirective = /** @class */ (function () {
    function IsNumberValidatorDirective() {
    }
    /**
     * @return {?}
     */
    IsNumberValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.isNumber;
    };
    /**
     * @param {?} c
     * @return {?}
     */
    IsNumberValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    IsNumberValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[isNumber][formControlName],[isNumber][formControl],[isNumber][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return IsNumberValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    return IsNumberValidatorDirective;
}());
export { IsNumberValidatorDirective };
if (false) {
    /** @type {?} */
    IsNumberValidatorDirective.prototype.validator;
}
var IsInRangeValidatorDirective = /** @class */ (function () {
    function IsInRangeValidatorDirective() {
    }
    /**
     * @return {?}
     */
    IsInRangeValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.isInRange(this.minValue, this.maxValue);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    IsInRangeValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    IsInRangeValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[isInRange][formControlName],[isInRange][formControl],[isInRange][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return IsInRangeValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    IsInRangeValidatorDirective.propDecorators = {
        minValue: [{ type: Input }],
        maxValue: [{ type: Input }]
    };
    return IsInRangeValidatorDirective;
}());
export { IsInRangeValidatorDirective };
if (false) {
    /** @type {?} */
    IsInRangeValidatorDirective.prototype.minValue;
    /** @type {?} */
    IsInRangeValidatorDirective.prototype.maxValue;
    /** @type {?} */
    IsInRangeValidatorDirective.prototype.validator;
}
var MaxValidatorDirective = /** @class */ (function () {
    function MaxValidatorDirective() {
    }
    /**
     * @return {?}
     */
    MaxValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.max(this.max);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    MaxValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    MaxValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[type=text][max][formControlName],input[type=text][max][formControl],input[type=text][max][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return MaxValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    MaxValidatorDirective.propDecorators = {
        max: [{ type: Input }]
    };
    return MaxValidatorDirective;
}());
export { MaxValidatorDirective };
if (false) {
    /** @type {?} */
    MaxValidatorDirective.prototype.max;
    /** @type {?} */
    MaxValidatorDirective.prototype.validator;
}
var MinValidatorDirective = /** @class */ (function () {
    function MinValidatorDirective() {
    }
    /**
     * @return {?}
     */
    MinValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = UniversalValidators.min(this.min);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    MinValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    MinValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'input[type=text][min][formControlName],input[type=text][min][formControl],input[type=text][min][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return MinValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    MinValidatorDirective.propDecorators = {
        min: [{ type: Input }]
    };
    return MinValidatorDirective;
}());
export { MinValidatorDirective };
if (false) {
    /** @type {?} */
    MinValidatorDirective.prototype.min;
    /** @type {?} */
    MinValidatorDirective.prototype.validator;
}
