// ng-zorro-antd/esm2015/tabs/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
export { NzTabBodyComponent } from './tab-body.component';
export { NzTabLabelDirective } from './tab-label.directive';
export { NzTabComponent } from './tab.component';
export { NzTabsInkBarDirective } from './tabs-ink-bar.directive';
export { NzTabsModule } from './tabs.module';
export { NzTabsNavComponent } from './tabs-nav.component';
export { NzTabSetComponent } from './tabset.component';
export { NzTabDirective } from './tab.directive';
export { NzTabLinkDirective } from './tab-link.directive';
export { NzTabChangeEvent } from './table.types';
