// ng-zorro-antd/esm5/button/button-group.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: button-group.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
var NzButtonGroupComponent = /** @class */ (function () {
    function NzButtonGroupComponent() {
        this.nzSize = 'default';
    }
    NzButtonGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-button-group',
                    exportAs: 'nzButtonGroup',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.ant-btn-group]': "true",
                        '[class.ant-btn-group-lg]': "nzSize === 'large'",
                        '[class.ant-btn-group-sm]': "nzSize === 'small'"
                    },
                    preserveWhitespaces: false,
                    template: "\n    <ng-content></ng-content>\n  "
                }] }
    ];
    NzButtonGroupComponent.propDecorators = {
        nzSize: [{ type: Input }]
    };
    return NzButtonGroupComponent;
}());
export { NzButtonGroupComponent };
if (false) {
    /** @type {?} */
    NzButtonGroupComponent.prototype.nzSize;
}
