// @ng-idle/keepalive/fesm2015/ng-idle-keepalive.js
import { EventEmitter, Injectable, NgZone, NgModule } from '@angular/core';
import { HttpRequest, HttpClient } from '@angular/common/http';
import { KeepaliveSvc, NgIdleModule } from '@ng-idle/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * An example of an injectable service.
 */
class Keepalive extends KeepaliveSvc {
    /*
       * Initializes a new instance of Keepalive
       * @param http - The HTTP service.
       */
    /**
     * @param {?} http
     * @param {?} zone
     */
    constructor(http, zone) {
        super();
        this.http = http;
        this.zone = zone;
        this.pingInterval = 10 * 60;
        /*
           * An event emitted when the service is pinging.
           */
        this.onPing = new EventEmitter();
        /*
           * An event emitted when the service has pinged an HTTP endpoint and received a response.
           */
        this.onPingResponse = new EventEmitter();
    }
    /*
       * Sets the string or Request that should be used when pinging.
       * @param url - The URL or Request object to use when pinging.
       * @return The current Request used when pinging.
       */
    /**
     * @template T
     * @param {?=} url
     * @return {?}
     */
    request(url) {
        if (typeof url === 'string') {
            this.pingRequest = new HttpRequest('GET', url);
        }
        else if (url instanceof HttpRequest) {
            this.pingRequest = url;
        }
        else if (url === null) {
            this.pingRequest = null;
        }
        return this.pingRequest;
    }
    /*
       * Sets the interval (in seconds) at which the ping operation will occur when start() is called.
       * @param seconds - The ping interval in seconds.
       * @return The current interval value.
       */
    /**
     * @param {?=} seconds
     * @return {?}
     */
    interval(seconds) {
        if (!isNaN(seconds) && seconds > 0) {
            this.pingInterval = seconds;
        }
        else if (!isNaN(seconds) && seconds <= 0) {
            throw new Error('Interval value must be greater than zero.');
        }
        return this.pingInterval;
    }
    /*
       * Immediately performs the ping operation. If a request has been set, an HTTP
       * request will be made and the response will be emitted via the
       * onPingResponse event.
       */
    /**
     * @return {?}
     */
    ping() {
        this.onPing.emit(null);
        if (this.pingRequest) {
            this.http.request(this.pingRequest).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                this.onPingResponse.emit(response);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            (error) => {
                this.onPingResponse.emit(error);
            }));
        }
    }
    /*
       * Starts pinging on an interval.
       */
    /**
     * @return {?}
     */
    start() {
        this.stop();
        this.zone.runOutsideAngular((/**
         * @return {?}
         */
        () => {
            this.pingHandle = setInterval((/**
             * @return {?}
             */
            () => {
                this.zone.run((/**
                 * @return {?}
                 */
                () => {
                    this.ping();
                }));
            }), this.pingInterval * 1000);
        }));
    }
    /*
       * Stops pinging on an interval.
       */
    /**
     * @return {?}
     */
    stop() {
        if (this.hasPingHandle()) {
            clearInterval(this.pingHandle);
            this.pingHandle = null;
        }
    }
    /*
       * Performs any cleanup tasks when Angular destroys the instance.
       */
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.stop();
    }
    /*
       * Returns whether or not the service will ping automatically at the specified interval.
       * @return True if the service will ping at the specified interval; otherwise, false.
       */
    /**
     * @return {?}
     */
    isRunning() {
        return this.hasPingHandle();
    }
    /**
     * @private
     * @return {?}
     */
    hasPingHandle() {
        return this.pingHandle !== null && typeof this.pingHandle !== 'undefined';
    }
}
Keepalive.decorators = [
    { type: Injectable }
];
/** @nocollapse */
Keepalive.ctorParameters = () => [
    { type: HttpClient },
    { type: NgZone }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NgIdleKeepaliveModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NgIdleKeepaliveModule,
            providers: [Keepalive, { provide: KeepaliveSvc, useExisting: Keepalive }]
        };
    }
}
NgIdleKeepaliveModule.decorators = [
    { type: NgModule, args: [{ imports: [NgIdleModule.forRoot()] },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { Keepalive, NgIdleKeepaliveModule };
