// ngx-validators/esm5/components/password/password.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS, Validators } from '@angular/forms';
import { PasswordValidators } from './password-validators';
var PasswordValidatorDirective = /** @class */ (function () {
    function PasswordValidatorDirective() {
        this.repeatCharacter = 4;
        this.alphabeticalCharacter = 1;
        this.digitCharacter = 1;
        this.lowercaseCharacter = 1;
        this.uppercaseCharacter = 1;
    }
    /**
     * @return {?}
     */
    PasswordValidatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.repeatCharacterValidator = PasswordValidators.repeatCharacterRegexRule(this.repeatCharacter);
        this.alphabeticalCharacterValidator = PasswordValidators.alphabeticalCharacterRule(this.alphabeticalCharacter);
        this.digitCharacterValidator = PasswordValidators.digitCharacterRule(this.digitCharacter);
        this.lowercaseCharacterValidator = PasswordValidators.lowercaseCharacterRule(this.lowercaseCharacter);
        this.uppercaseCharacterValidator = PasswordValidators.uppercaseCharacterRule(this.uppercaseCharacter);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    PasswordValidatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        /** @type {?} */
        var compose = Validators.compose([this.repeatCharacterValidator,
            this.digitCharacterValidator, this.alphabeticalCharacterValidator,
            this.lowercaseCharacterValidator, this.uppercaseCharacterValidator]);
        return compose(c);
    };
    PasswordValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[password][formControlName],[password][formControl],[password][ngModel]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return PasswordValidatorDirective; }),
                            multi: true
                        }]
                },] }
    ];
    PasswordValidatorDirective.propDecorators = {
        repeatCharacter: [{ type: Input }],
        alphabeticalCharacter: [{ type: Input }],
        digitCharacter: [{ type: Input }],
        lowercaseCharacter: [{ type: Input }],
        uppercaseCharacter: [{ type: Input }]
    };
    return PasswordValidatorDirective;
}());
export { PasswordValidatorDirective };
if (false) {
    /** @type {?} */
    PasswordValidatorDirective.prototype.repeatCharacter;
    /** @type {?} */
    PasswordValidatorDirective.prototype.alphabeticalCharacter;
    /** @type {?} */
    PasswordValidatorDirective.prototype.digitCharacter;
    /** @type {?} */
    PasswordValidatorDirective.prototype.lowercaseCharacter;
    /** @type {?} */
    PasswordValidatorDirective.prototype.uppercaseCharacter;
    /** @type {?} */
    PasswordValidatorDirective.prototype.repeatCharacterValidator;
    /** @type {?} */
    PasswordValidatorDirective.prototype.alphabeticalCharacterValidator;
    /** @type {?} */
    PasswordValidatorDirective.prototype.digitCharacterValidator;
    /** @type {?} */
    PasswordValidatorDirective.prototype.lowercaseCharacterValidator;
    /** @type {?} */
    PasswordValidatorDirective.prototype.uppercaseCharacterValidator;
}
