// ngx-validators/esm5/components/equal-to/equal-to.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { delay } from 'rxjs/operators';
var EqualToDirective = /** @class */ (function () {
    function EqualToDirective() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    EqualToDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        /** @type {?} */
        var otherControl = typeof this.equalTo === 'string' ? c.parent.get(this.equalTo) : this.equalTo;
        if (!this.subscription) {
            this.subscription = otherControl.valueChanges
                .pipe(delay(1)).subscribe(function () {
                c.updateValueAndValidity();
            });
        }
        return c.value !== otherControl.value ? { notEqualTo: true } : null;
    };
    /**
     * @return {?}
     */
    EqualToDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscription.unsubscribe();
    };
    EqualToDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[equalTo][ngModel], [equalTo][formControlName], [equalTo][formControl]',
                    providers: [{
                            provide: NG_VALIDATORS,
                            // tslint:disable-next-line:no-forward-ref
                            useExisting: forwardRef(function () { return EqualToDirective; }),
                            multi: true
                        }]
                },] }
    ];
    EqualToDirective.propDecorators = {
        equalTo: [{ type: Input }]
    };
    return EqualToDirective;
}());
export { EqualToDirective };
if (false) {
    /** @type {?} */
    EqualToDirective.prototype.equalTo;
    /** @type {?} */
    EqualToDirective.prototype.subscription;
}
