// @ng-idle/core/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of core
 */
export { AutoResume, Idle } from './lib/idle';
export { InterruptArgs } from './lib/interruptargs';
export { InterruptSource } from './lib/interruptsource';
export { EventTargetInterruptSource } from './lib/eventtargetinterruptsource';
export { DocumentInterruptSource } from './lib/documentinterruptsource';
export { WindowInterruptSource } from './lib/windowinterruptsource';
export { StorageInterruptSource } from './lib/storageinterruptsource';
export { KeepaliveSvc } from './lib/keepalivesvc';
export { IdleExpiry } from './lib/idleexpiry';
export { SimpleExpiry } from './lib/simpleexpiry';
export { LocalStorage } from './lib/localstorage';
export { LocalStorageExpiry } from './lib/localstorageexpiry';
export { NgIdleModule } from './lib/module';
export { createDefaultInterruptSources, DEFAULT_INTERRUPTSOURCES } from './lib/defaultinterruptsources';
