// ng-zorro-antd/esm5/result/result.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: result.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
/** @type {?} */
var IconMap = {
    success: 'check-circle',
    error: 'close-circle',
    info: 'exclamation-circle',
    warning: 'warning'
};
/** @type {?} */
var ExceptionStatus = ['404', '500', '403'];
var NzResultComponent = /** @class */ (function () {
    function NzResultComponent() {
        this.nzStatus = 'info';
        this.isException = false;
    }
    /**
     * @return {?}
     */
    NzResultComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.setStatusIcon();
    };
    /**
     * @private
     * @return {?}
     */
    NzResultComponent.prototype.setStatusIcon = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var icon = this.nzIcon;
        this.isException = ExceptionStatus.indexOf(this.nzStatus) !== -1;
        this.icon = icon
            ? typeof icon === 'string'
                ? IconMap[(/** @type {?} */ (icon))] || icon
                : icon
            : this.isException
                ? undefined
                : IconMap[(/** @type {?} */ (this.nzStatus))];
    };
    NzResultComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    selector: 'nz-result',
                    exportAs: 'nzResult',
                    template: "\n    <div class=\"ant-result-icon\">\n      <ng-container *ngIf=\"!isException; else exceptionTpl\">\n        <ng-container *ngIf=\"icon\">\n          <ng-container *nzStringTemplateOutlet=\"icon\">\n            <i nz-icon [nzType]=\"icon\" nzTheme=\"fill\"></i>\n          </ng-container>\n        </ng-container>\n        <ng-content *ngIf=\"!icon\" select=\"[nz-result-icon]\"></ng-content>\n      </ng-container>\n    </div>\n    <ng-container *ngIf=\"nzTitle\">\n      <div class=\"ant-result-title\" *nzStringTemplateOutlet=\"nzTitle\">\n        {{ nzTitle }}\n      </div>\n    </ng-container>\n    <ng-content *ngIf=\"!nzTitle\" select=\"div[nz-result-title]\"></ng-content>\n    <ng-container *ngIf=\"nzSubTitle\">\n      <div class=\"ant-result-subtitle\" *nzStringTemplateOutlet=\"nzSubTitle\">\n        {{ nzSubTitle }}\n      </div>\n    </ng-container>\n    <ng-content *ngIf=\"!nzSubTitle\" select=\"div[nz-result-subtitle]\"></ng-content>\n    <ng-content select=\"nz-result-content, [nz-result-content]\"></ng-content>\n    <div class=\"ant-result-extra\" *ngIf=\"nzExtra\">\n      <ng-container *nzStringTemplateOutlet=\"nzExtra\">\n        {{ nzExtra }}\n      </ng-container>\n    </div>\n    <ng-content *ngIf=\"!nzExtra\" select=\"div[nz-result-extra]\"></ng-content>\n\n    <ng-template #exceptionTpl>\n      <ng-container [ngSwitch]=\"nzStatus\">\n        <nz-result-not-found *ngSwitchCase=\"'404'\"></nz-result-not-found>\n        <nz-result-server-error *ngSwitchCase=\"'500'\"></nz-result-server-error>\n        <nz-result-unauthorized *ngSwitchCase=\"'403'\"></nz-result-unauthorized>\n      </ng-container>\n    </ng-template>\n  ",
                    host: {
                        '[class.ant-result]': 'true',
                        '[class.ant-result-success]': "nzStatus === 'success'",
                        '[class.ant-result-error]': "nzStatus === 'error'",
                        '[class.ant-result-info]': "nzStatus === 'info'",
                        '[class.ant-result-warning]': "nzStatus === 'warning'"
                    }
                }] }
    ];
    /** @nocollapse */
    NzResultComponent.ctorParameters = function () { return []; };
    NzResultComponent.propDecorators = {
        nzIcon: [{ type: Input }],
        nzTitle: [{ type: Input }],
        nzStatus: [{ type: Input }],
        nzSubTitle: [{ type: Input }],
        nzExtra: [{ type: Input }]
    };
    return NzResultComponent;
}());
export { NzResultComponent };
if (false) {
    /** @type {?} */
    NzResultComponent.prototype.nzIcon;
    /** @type {?} */
    NzResultComponent.prototype.nzTitle;
    /** @type {?} */
    NzResultComponent.prototype.nzStatus;
    /** @type {?} */
    NzResultComponent.prototype.nzSubTitle;
    /** @type {?} */
    NzResultComponent.prototype.nzExtra;
    /** @type {?} */
    NzResultComponent.prototype.icon;
    /** @type {?} */
    NzResultComponent.prototype.isException;
}
