// @ng-idle/core/esm5/lib/interruptsource.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
/*
 * A base for classes that act as a source for interrupts.
 */
/**
 * @abstract
 */
var /*
 * A base for classes that act as a source for interrupts.
 */
/**
 * @abstract
 */
InterruptSource = /** @class */ (function () {
    function InterruptSource(attachFn, detachFn) {
        this.attachFn = attachFn;
        this.detachFn = detachFn;
        this.isAttached = false;
        this.onInterrupt = new EventEmitter();
    }
    /*
     * Attaches to the specified events on the specified source.
     */
    /*
       * Attaches to the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    InterruptSource.prototype.attach = /*
       * Attaches to the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        // If the current zone is the 'angular' zone (a.k.a. NgZone) then re-enter this method in its parent zone
        // The parent zone is usually the '<root>' zone but it can also be 'long-stack-trace-zone' in debug mode
        // In tests, the current zone is typically a 'ProxyZone' created by async/fakeAsync (from @angular/core/testing)
        if (Zone.current.get('isAngularZone') === true) {
            Zone.current.parent.run((/**
             * @return {?}
             */
            function () { return _this.attach(); }));
            return;
        }
        if (!this.isAttached && this.attachFn) {
            this.attachFn(this);
        }
        this.isAttached = true;
    };
    /*
     * Detaches from the specified events on the specified source.
     */
    /*
       * Detaches from the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    InterruptSource.prototype.detach = /*
       * Detaches from the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    function () {
        if (this.isAttached && this.detachFn) {
            this.detachFn(this);
        }
        this.isAttached = false;
    };
    return InterruptSource;
}());
/*
 * A base for classes that act as a source for interrupts.
 */
/**
 * @abstract
 */
export { InterruptSource };
if (false) {
    /** @type {?} */
    InterruptSource.prototype.isAttached;
    /** @type {?} */
    InterruptSource.prototype.onInterrupt;
    /**
     * @type {?}
     * @protected
     */
    InterruptSource.prototype.attachFn;
    /**
     * @type {?}
     * @protected
     */
    InterruptSource.prototype.detachFn;
}
