// @ng-idle/core/esm5/lib/interrupt.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * A class for managing an interrupt from an interrupt source.
 */
var /*
 * A class for managing an interrupt from an interrupt source.
 */
Interrupt = /** @class */ (function () {
    function Interrupt(source) {
        this.source = source;
    }
    /*
     * Subscribes to the interrupt using the specified function.
     * @param fn - The subscription function.
     */
    /*
       * Subscribes to the interrupt using the specified function.
       * @param fn - The subscription function.
       */
    /**
     * @param {?} fn
     * @return {?}
     */
    Interrupt.prototype.subscribe = /*
       * Subscribes to the interrupt using the specified function.
       * @param fn - The subscription function.
       */
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.sub = this.source.onInterrupt.subscribe(fn);
    };
    /*
     * Unsubscribes the interrupt.
     */
    /*
       * Unsubscribes the interrupt.
       */
    /**
     * @return {?}
     */
    Interrupt.prototype.unsubscribe = /*
       * Unsubscribes the interrupt.
       */
    /**
     * @return {?}
     */
    function () {
        this.sub.unsubscribe();
        this.sub = null;
    };
    /*
     * Keeps the subscription but resumes interrupt events.
     */
    /*
       * Keeps the subscription but resumes interrupt events.
       */
    /**
     * @return {?}
     */
    Interrupt.prototype.resume = /*
       * Keeps the subscription but resumes interrupt events.
       */
    /**
     * @return {?}
     */
    function () {
        this.source.attach();
    };
    /*
     * Keeps the subscription but pauses interrupt events.
     */
    /*
       * Keeps the subscription but pauses interrupt events.
       */
    /**
     * @return {?}
     */
    Interrupt.prototype.pause = /*
       * Keeps the subscription but pauses interrupt events.
       */
    /**
     * @return {?}
     */
    function () {
        this.source.detach();
    };
    return Interrupt;
}());
/*
 * A class for managing an interrupt from an interrupt source.
 */
export { Interrupt };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Interrupt.prototype.sub;
    /** @type {?} */
    Interrupt.prototype.source;
}
