// ng-zorro-antd/esm5/select/select-placeholder.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: select-placeholder.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
var NzSelectPlaceholderComponent = /** @class */ (function () {
    function NzSelectPlaceholderComponent() {
        this.placeholder = null;
    }
    NzSelectPlaceholderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-select-placeholder',
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-container *nzStringTemplateOutlet=\"placeholder\">\n      {{ placeholder }}\n    </ng-container>\n  ",
                    host: {
                        '[class.ant-select-selection-placeholder]': 'true'
                    }
                }] }
    ];
    NzSelectPlaceholderComponent.propDecorators = {
        placeholder: [{ type: Input }]
    };
    return NzSelectPlaceholderComponent;
}());
export { NzSelectPlaceholderComponent };
if (false) {
    /** @type {?} */
    NzSelectPlaceholderComponent.prototype.placeholder;
}
