// ng-zorro-antd/esm5/radio/radio.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: radio.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Injectable } from '@angular/core';
import { ReplaySubject, Subject } from 'rxjs';
var NzRadioService = /** @class */ (function () {
    function NzRadioService() {
        this.selected$ = new ReplaySubject(1);
        this.touched$ = new Subject();
        this.disabled$ = new ReplaySubject(1);
        this.name$ = new ReplaySubject(1);
    }
    /**
     * @return {?}
     */
    NzRadioService.prototype.touch = /**
     * @return {?}
     */
    function () {
        this.touched$.next();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NzRadioService.prototype.select = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.selected$.next(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NzRadioService.prototype.setDisabled = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.disabled$.next(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NzRadioService.prototype.setName = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.name$.next(value);
    };
    NzRadioService.decorators = [
        { type: Injectable }
    ];
    return NzRadioService;
}());
export { NzRadioService };
if (false) {
    /** @type {?} */
    NzRadioService.prototype.selected$;
    /** @type {?} */
    NzRadioService.prototype.touched$;
    /** @type {?} */
    NzRadioService.prototype.disabled$;
    /** @type {?} */
    NzRadioService.prototype.name$;
}
