// ng-zorro-antd/esm2015/date-picker/lib/interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
/**
 * @record
 */
export function PanelSelector() { }
if (false) {
    /** @type {?} */
    PanelSelector.prototype.className;
    /** @type {?|undefined} */
    PanelSelector.prototype.title;
    /** @type {?} */
    PanelSelector.prototype.label;
    /**
     * @return {?}
     */
    PanelSelector.prototype.onClick = function () { };
}
/**
 * @record
 */
export function DateCell() { }
if (false) {
    /** @type {?} */
    DateCell.prototype.value;
    /** @type {?} */
    DateCell.prototype.content;
    /** @type {?} */
    DateCell.prototype.isDisabled;
    /** @type {?} */
    DateCell.prototype.isSelected;
    /** @type {?|undefined} */
    DateCell.prototype.label;
    /** @type {?|undefined} */
    DateCell.prototype.title;
    /** @type {?|undefined} */
    DateCell.prototype.cellRender;
    /** @type {?|undefined} */
    DateCell.prototype.fullCellRender;
    /** @type {?|undefined} */
    DateCell.prototype.isToday;
    /** @type {?|undefined} */
    DateCell.prototype.classMap;
    /**
     * @return {?}
     */
    DateCell.prototype.onClick = function () { };
    /**
     * @return {?}
     */
    DateCell.prototype.onMouseEnter = function () { };
}
/**
 * @record
 */
export function DateBodyRow() { }
if (false) {
    /** @type {?} */
    DateBodyRow.prototype.dateCells;
    /** @type {?|undefined} */
    DateBodyRow.prototype.isCurrent;
    /** @type {?|undefined} */
    DateBodyRow.prototype.isActive;
    /** @type {?|undefined} */
    DateBodyRow.prototype.weekNum;
    /** @type {?|undefined} */
    DateBodyRow.prototype.year;
    /** @type {?|undefined} */
    DateBodyRow.prototype.classMap;
}
/**
 * @record
 */
export function DayCell() { }
if (false) {
    /** @type {?|undefined} */
    DayCell.prototype.isSelectedStartDate;
    /** @type {?|undefined} */
    DayCell.prototype.isSelectedEndDate;
    /** @type {?|undefined} */
    DayCell.prototype.isHoverStartDate;
    /** @type {?|undefined} */
    DayCell.prototype.isHoverEndDate;
    /** @type {?|undefined} */
    DayCell.prototype.isInHoverRange;
    /** @type {?|undefined} */
    DayCell.prototype.isInSelectedRange;
    /** @type {?|undefined} */
    DayCell.prototype.isLastDayOfMonth;
    /** @type {?|undefined} */
    DayCell.prototype.isFirstDayOfMonth;
    /** @type {?|undefined} */
    DayCell.prototype.isStartSingle;
    /** @type {?|undefined} */
    DayCell.prototype.isEndSingle;
}
/**
 * @record
 */
export function DecadeCell() { }
if (false) {
    /** @type {?|undefined} */
    DecadeCell.prototype.isBiggerThanEnd;
    /** @type {?|undefined} */
    DecadeCell.prototype.isLowerThanStart;
}
/**
 * @record
 */
export function YearCell() { }
if (false) {
    /** @type {?|undefined} */
    YearCell.prototype.isSameDecade;
}
