// ng-zorro-antd/esm5/layout/layout.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: layout.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, ContentChildren, QueryList, ViewEncapsulation } from '@angular/core';
import { NzSiderComponent } from './sider.component';
var NzLayoutComponent = /** @class */ (function () {
    function NzLayoutComponent() {
    }
    NzLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-layout',
                    exportAs: 'nzLayout',
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        '[class.ant-layout-has-sider]': 'listOfNzSiderComponent.length > 0',
                        '[class.ant-layout]': 'true'
                    }
                }] }
    ];
    NzLayoutComponent.propDecorators = {
        listOfNzSiderComponent: [{ type: ContentChildren, args: [NzSiderComponent,] }]
    };
    return NzLayoutComponent;
}());
export { NzLayoutComponent };
if (false) {
    /** @type {?} */
    NzLayoutComponent.prototype.listOfNzSiderComponent;
}
