// ng-zorro-antd/esm5/modal/modal-legacy-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: modal-legacy-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
/**
 * @abstract
 * @template T, R
 */
var /**
 * @abstract
 * @template T, R
 */
NzModalLegacyAPI = /** @class */ (function () {
    function NzModalLegacyAPI() {
    }
    return NzModalLegacyAPI;
}());
/**
 * @abstract
 * @template T, R
 */
export { NzModalLegacyAPI };
if (false) {
    /** @type {?} */
    NzModalLegacyAPI.prototype.afterOpen;
    /** @type {?} */
    NzModalLegacyAPI.prototype.afterClose;
    /**
     * @abstract
     * @return {?}
     */
    NzModalLegacyAPI.prototype.open = function () { };
    /**
     * @abstract
     * @param {?=} result
     * @return {?}
     */
    NzModalLegacyAPI.prototype.close = function (result) { };
    /**
     * @abstract
     * @param {?=} result
     * @return {?}
     */
    NzModalLegacyAPI.prototype.destroy = function (result) { };
    /**
     * Trigger the nzOnOk/nzOnCancel by manual
     * @abstract
     * @return {?}
     */
    NzModalLegacyAPI.prototype.triggerOk = function () { };
    /**
     * @abstract
     * @return {?}
     */
    NzModalLegacyAPI.prototype.triggerCancel = function () { };
    /**
     * Return the component instance of nzContent when specify nzContent as a Component
     * @abstract
     * @return {?}
     */
    NzModalLegacyAPI.prototype.getContentComponent = function () { };
    /**
     * Get the dom element of this Modal
     * @abstract
     * @return {?}
     */
    NzModalLegacyAPI.prototype.getElement = function () { };
}
