// ng-zorro-antd/esm5/space/space-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: space-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, ElementRef, Renderer2 } from '@angular/core';
/** @type {?} */
var spaceSize = {
    small: 8,
    middle: 16,
    large: 24
};
var NzSpaceItemComponent = /** @class */ (function () {
    function NzSpaceItemComponent(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
    }
    /**
     * @param {?} direction
     * @param {?} size
     * @return {?}
     */
    NzSpaceItemComponent.prototype.setDirectionAndSize = /**
     * @param {?} direction
     * @param {?} size
     * @return {?}
     */
    function (direction, size) {
        /** @type {?} */
        var marginSize = typeof size === 'string' ? spaceSize[size] : size;
        if (direction === 'horizontal') {
            this.renderer.removeStyle(this.elementRef.nativeElement, 'margin-bottom');
            this.renderer.setStyle(this.elementRef.nativeElement, 'margin-right', marginSize + "px");
        }
        else {
            this.renderer.removeStyle(this.elementRef.nativeElement, 'margin-right');
            this.renderer.setStyle(this.elementRef.nativeElement, 'margin-bottom', marginSize + "px");
        }
    };
    /**
     * @return {?}
     */
    NzSpaceItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    NzSpaceItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-space-item, [nz-space-item]',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-content></ng-content>\n  ",
                    host: {
                        class: 'ant-space-item'
                    }
                }] }
    ];
    /** @nocollapse */
    NzSpaceItemComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    return NzSpaceItemComponent;
}());
export { NzSpaceItemComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NzSpaceItemComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    NzSpaceItemComponent.prototype.elementRef;
}
