// ng-zorro-antd/esm5/dropdown/dropdown-button.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: dropdown-button.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive, ElementRef, Host, Optional, Renderer2 } from '@angular/core';
import { NzButtonGroupComponent } from 'ng-zorro-antd/button';
var NzDropdownButtonDirective = /** @class */ (function () {
    function NzDropdownButtonDirective(renderer, nzButtonGroupComponent, elementRef) {
        this.renderer = renderer;
        this.nzButtonGroupComponent = nzButtonGroupComponent;
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    NzDropdownButtonDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var parentElement = this.renderer.parentNode(this.elementRef.nativeElement);
        if (this.nzButtonGroupComponent && parentElement) {
            this.renderer.addClass(parentElement, 'ant-dropdown-button');
        }
    };
    NzDropdownButtonDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[nz-button][nz-dropdown]'
                },] }
    ];
    /** @nocollapse */
    NzDropdownButtonDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: NzButtonGroupComponent, decorators: [{ type: Host }, { type: Optional }] },
        { type: ElementRef }
    ]; };
    return NzDropdownButtonDirective;
}());
export { NzDropdownButtonDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NzDropdownButtonDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    NzDropdownButtonDirective.prototype.nzButtonGroupComponent;
    /**
     * @type {?}
     * @private
     */
    NzDropdownButtonDirective.prototype.elementRef;
}
