// ng-zorro-antd/esm2015/slider/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
export { NzSliderComponent } from './slider.component';
export { NzSliderService as ɵNzSliderService } from './slider.service';
export { NzSliderModule } from './slider.module';
export { NzSliderHandleComponent as ɵNzSliderHandleComponent } from './handle.component';
export { NzSliderMarksComponent as ɵNzSliderMarksComponent } from './marks.component';
export { NzSliderStepComponent as ɵNzSliderStepComponent } from './step.component';
export { NzSliderTrackComponent as ɵNzSliderTrackComponent } from './track.component';
export { NzMarks } from './typings';
