// @ng-idle/core/esm2015/lib/alternativestorage.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Represents an alternative storage for browser that doesn't support localstorage. (i.e. Safari in
 * private mode)
 * @implements Storage
 */
export class AlternativeStorage {
    constructor() {
        this.storageMap = {};
    }
    /*
       * Returns an integer representing the number of data items stored in the storageMap object.
       */
    /**
     * @return {?}
     */
    get length() {
        return Object.keys(this.storageMap).length;
    }
    /*
       * Remove all keys out of the storage.
       */
    /**
     * @return {?}
     */
    clear() {
        this.storageMap = {};
    }
    /*
       * Return the key's value
       *
       * @param key - name of the key to retrieve the value of.
       * @return The key's value
       */
    /**
     * @param {?} key
     * @return {?}
     */
    getItem(key) {
        if (typeof this.storageMap[key] !== 'undefined') {
            return this.storageMap[key];
        }
        return null;
    }
    /*
       * Return the nth key in the storage
       *
       * @param index - the number of the key you want to get the name of.
       * @return The name of the key.
       */
    /**
     * @param {?} index
     * @return {?}
     */
    key(index) {
        return Object.keys(this.storageMap)[index] || null;
    }
    /*
       * Remove a key from the storage.
       *
       * @param key - the name of the key you want to remove.
       */
    /**
     * @param {?} key
     * @return {?}
     */
    removeItem(key) {
        this.storageMap[key] = undefined;
    }
    /*
       * Add a key to the storage, or update a key's value if it already exists.
       *
       * @param key - the name of the key.
       * @param value - the value you want to give to the key.
       */
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setItem(key, value) {
        this.storageMap[key] = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AlternativeStorage.prototype.storageMap;
    /* Skipping unhandled member: [key: string]: any;*/
    /* Skipping unhandled member: [index: number]: string;*/
}
