// @ng-idle/core/esm2015/lib/eventtargetinterruptsource.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subscription, fromEvent, merge } from 'rxjs';
import { filter, throttleTime } from 'rxjs/operators';
import { InterruptArgs } from './interruptargs';
import { InterruptSource } from './interruptsource';
/**
 * Options for EventTargetInterruptSource
 * @record
 */
export function EventTargetInterruptOptions() { }
if (false) {
    /**
     * The number of milliseconds to throttle the events coming from the target.
     * @type {?|undefined}
     */
    EventTargetInterruptOptions.prototype.throttleDelay;
    /**
     * Whether or not to use passive event listeners.
     * Note: you need to detect if the browser supports passive listeners, and only set this to true if it does.
     * @type {?|undefined}
     */
    EventTargetInterruptOptions.prototype.passive;
}
/** @type {?} */
const defaultThrottleDelay = 500;
/*
 * An interrupt source on an EventTarget object, such as a Window or HTMLElement.
 */
export class EventTargetInterruptSource extends InterruptSource {
    /**
     * @param {?} target
     * @param {?} events
     * @param {?=} options
     */
    constructor(target, events, options) {
        super(null, null);
        this.target = target;
        this.events = events;
        this.eventSubscription = new Subscription();
        if (typeof options === 'number') {
            options = { throttleDelay: options, passive: false };
        }
        options = options || {
            throttleDelay: defaultThrottleDelay,
            passive: false
        };
        if (options.throttleDelay === undefined || options.throttleDelay === null) {
            options.throttleDelay = defaultThrottleDelay;
        }
        this.throttleDelay = options.throttleDelay;
        this.passive = !!options.passive;
        /** @type {?} */
        const opts = this.passive ? { passive: true } : null;
        /** @type {?} */
        const fromEvents = events
            .split(' ')
            .map((/**
         * @param {?} eventName
         * @return {?}
         */
        eventName => fromEvent(target, eventName, opts)));
        this.eventSrc = merge(...fromEvents);
        this.eventSrc = this.eventSrc.pipe(filter((/**
         * @param {?} innerArgs
         * @return {?}
         */
        innerArgs => !this.filterEvent(innerArgs))));
        if (this.throttleDelay > 0) {
            this.eventSrc = this.eventSrc.pipe(throttleTime(this.throttleDelay));
        }
        /** @type {?} */
        const handler = (/**
         * @param {?} innerArgs
         * @return {?}
         */
        (innerArgs) => this.onInterrupt.emit(new InterruptArgs(this, innerArgs)));
        this.attachFn = (/**
         * @return {?}
         */
        () => (this.eventSubscription = this.eventSrc.subscribe(handler)));
        this.detachFn = (/**
         * @return {?}
         */
        () => this.eventSubscription.unsubscribe());
    }
    /*
       * Checks to see if the event should be filtered. Always returns false unless overriden.
       * @param event - The original event object.
       * @return True if the event should be filtered (don't cause an interrupt); otherwise, false.
       */
    /**
     * @protected
     * @param {?} event
     * @return {?}
     */
    filterEvent(event) {
        return false;
    }
    /**
     * Returns the current options being used.
     * @return {?} The current option values.
     */
    get options() {
        return { throttleDelay: this.throttleDelay, passive: this.passive };
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventTargetInterruptSource.prototype.eventSrc;
    /**
     * @type {?}
     * @private
     */
    EventTargetInterruptSource.prototype.eventSubscription;
    /**
     * @type {?}
     * @protected
     */
    EventTargetInterruptSource.prototype.throttleDelay;
    /**
     * @type {?}
     * @protected
     */
    EventTargetInterruptSource.prototype.passive;
    /**
     * @type {?}
     * @protected
     */
    EventTargetInterruptSource.prototype.target;
    /**
     * @type {?}
     * @protected
     */
    EventTargetInterruptSource.prototype.events;
}
