// @ng-idle/core/esm2015/lib/idleexpiry.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Represents a base class for types that provide expiry detection for the Idle service.
 */
/**
 * @abstract
 */
export class IdleExpiry {
    constructor() {
        this.idValue = new Date();
        this.idlingValue = false;
    }
    /*
       * Gets or sets a unique ID for the window
       * @param id - The id.
       * @return The current id.
       */
    /**
     * @param {?=} value
     * @return {?}
     */
    id(value) {
        if (value !== void 0) {
            if (!value) {
                throw new Error('A value must be specified for the ID.');
            }
            this.idValue = value;
        }
        return this.idValue;
    }
    /*
       * Gets or sets the idling value.
       * @param value - The value to set.
       * @return The idling value.
       */
    /**
     * @param {?=} value
     * @return {?}
     */
    idling(value) {
        if (value !== void 0) {
            this.idlingValue = value;
        }
        return this.idlingValue;
    }
    /*
       * Returns the current Date.
       * @return The current Date.
       */
    /**
     * @return {?}
     */
    now() {
        /* istanbul ignore next */
        return new Date();
    }
    /*
       * Returns whether or not it is expired.
       * @return True if expired; otherwise, false.
       */
    /**
     * @return {?}
     */
    isExpired() {
        /** @type {?} */
        const expiry = this.last();
        return expiry != null && expiry <= this.now();
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    IdleExpiry.prototype.idValue;
    /**
     * @type {?}
     * @protected
     */
    IdleExpiry.prototype.idlingValue;
    /**
     * @abstract
     * @param {?=} value
     * @return {?}
     */
    IdleExpiry.prototype.last = function (value) { };
}
