// @ng-idle/core/esm2015/lib/interruptsource.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
/*
 * A base for classes that act as a source for interrupts.
 */
/**
 * @abstract
 */
export class InterruptSource {
    /**
     * @param {?=} attachFn
     * @param {?=} detachFn
     */
    constructor(attachFn, detachFn) {
        this.attachFn = attachFn;
        this.detachFn = detachFn;
        this.isAttached = false;
        this.onInterrupt = new EventEmitter();
    }
    /*
       * Attaches to the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    attach() {
        // If the current zone is the 'angular' zone (a.k.a. NgZone) then re-enter this method in its parent zone
        // The parent zone is usually the '<root>' zone but it can also be 'long-stack-trace-zone' in debug mode
        // In tests, the current zone is typically a 'ProxyZone' created by async/fakeAsync (from @angular/core/testing)
        if (Zone.current.get('isAngularZone') === true) {
            Zone.current.parent.run((/**
             * @return {?}
             */
            () => this.attach()));
            return;
        }
        if (!this.isAttached && this.attachFn) {
            this.attachFn(this);
        }
        this.isAttached = true;
    }
    /*
       * Detaches from the specified events on the specified source.
       */
    /**
     * @return {?}
     */
    detach() {
        if (this.isAttached && this.detachFn) {
            this.detachFn(this);
        }
        this.isAttached = false;
    }
}
if (false) {
    /** @type {?} */
    InterruptSource.prototype.isAttached;
    /** @type {?} */
    InterruptSource.prototype.onInterrupt;
    /**
     * @type {?}
     * @protected
     */
    InterruptSource.prototype.attachFn;
    /**
     * @type {?}
     * @protected
     */
    InterruptSource.prototype.detachFn;
}
