// @ng-idle/core/esm2015/lib/interrupt.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * A class for managing an interrupt from an interrupt source.
 */
export class Interrupt {
    /**
     * @param {?} source
     */
    constructor(source) {
        this.source = source;
    }
    /*
       * Subscribes to the interrupt using the specified function.
       * @param fn - The subscription function.
       */
    /**
     * @param {?} fn
     * @return {?}
     */
    subscribe(fn) {
        this.sub = this.source.onInterrupt.subscribe(fn);
    }
    /*
       * Unsubscribes the interrupt.
       */
    /**
     * @return {?}
     */
    unsubscribe() {
        this.sub.unsubscribe();
        this.sub = null;
    }
    /*
       * Keeps the subscription but resumes interrupt events.
       */
    /**
     * @return {?}
     */
    resume() {
        this.source.attach();
    }
    /*
       * Keeps the subscription but pauses interrupt events.
       */
    /**
     * @return {?}
     */
    pause() {
        this.source.detach();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Interrupt.prototype.sub;
    /** @type {?} */
    Interrupt.prototype.source;
}
